sprocs/assessment_instances_select_for_auto_finish.sql

```diff
@@ -15,12 +18,25 @@ BEGIN
             assessment_instances AS ai
             JOIN assessments AS a ON (a.id = ai.assessment_id)
         WHERE
-            ai.open = true
-            AND a.type = 'Exam'
-            AND ai.auto_close
+            a.type = 'Exam'
+            AND (
+                (ai.open AND ai.auto_close)
+                OR
+                (ai.open = false AND ai.grading_needed)
+            )
     LOOP
         assessment_instance_id := assessment_instance.id;
 
+        -- Only mark this assessment as needing to be closed if it's still open.
+        close_assessment := assessment_instance.open;
+
+        -- First check if the exam is closed and in need of grading. We can
+        -- immediately append it to the results if that's the case.
+        IF assessment_instance.open = false AND assessment_instance.grading_needed THEN
+            RETURN NEXT;
+            CONTINUE;
```

Comment:
I had to add `CONTINUE` here - TIL that `RETURN NEXT` does not in fact stop execution of the current loop iteration!

Reply:
Good catch! I knew that at some point, but apparently not when writing that code 😂

Reply:
This is my code, FWIW 😉

Reply:
Haha, nice 😄